
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      <link rel="icon" href="../../assets/images/SIB_logo.svg">
      <meta name="generator" content="mkdocs-1.2.2, mkdocs-material-7.2.5">
    
    
      
        <title>Introduction scRNAseq - Single cell transcriptomics</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.be71726b.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.3f5d1f46.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    
      


    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    <script>function __prefix(e){return new URL("../..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#learning-outcomes" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Single cell transcriptomics" class="md-header__button md-logo" aria-label="Single cell transcriptomics" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Single cell transcriptomics
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Introduction scRNAseq
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        
<a href="https://github.com/sib-swiss/single-cell-training/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/single-cell-training
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Single cell transcriptomics" class="md-nav__button md-logo" aria-label="Single cell transcriptomics" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    Single cell transcriptomics
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/sib-swiss/single-cell-training/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/single-cell-training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../precourse/" class="md-nav__link">
        Precourse preparations
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../course_schedule/" class="md-nav__link">
        Course schedule
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" checked>
      
      <label class="md-nav__link" for="__nav_4">
        Day 1
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Day 1" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Day 1
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../general_introduction/" class="md-nav__link">
        General introduction
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../setup/" class="md-nav__link">
        Setup
      </a>
    </li>
  

          
            
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Introduction scRNAseq
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Introduction scRNAseq
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    Learning outcomes
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    Material
  </a>
  
    <nav class="md-nav" aria-label="Material">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#running-cellranger-count" class="md-nav__link">
    Running cellranger count
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#loading-scrnaseq-data" class="md-nav__link">
    Loading scRNAseq data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#the-seurat-object" class="md-nav__link">
    The Seurat object
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../quality_control/" class="md-nav__link">
        Quality control
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../normalization_scaling/" class="md-nav__link">
        Normalization and scaling
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" >
      
      <label class="md-nav__link" for="__nav_5">
        Day 2
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Day 2" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          Day 2
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/dimensionality_reduction/" class="md-nav__link">
        Dimensionality reduction
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/integration/" class="md-nav__link">
        Integration
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/clustering/" class="md-nav__link">
        Clustering
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/cell_annotation/" class="md-nav__link">
        Cell annotation
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6" type="checkbox" id="__nav_6" >
      
      <label class="md-nav__link" for="__nav_6">
        Day 3
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Day 3" data-md-level="1">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          Day 3
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../day3/differential_gene_expression/" class="md-nav__link">
        Differential gene expression
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../day3/enrichment_analysis/" class="md-nav__link">
        Enrichment analysis
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../day3/trajectory_analysis/" class="md-nav__link">
        Trajectory analysis
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    Learning outcomes
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    Material
  </a>
  
    <nav class="md-nav" aria-label="Material">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#running-cellranger-count" class="md-nav__link">
    Running cellranger count
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#loading-scrnaseq-data" class="md-nav__link">
    Loading scRNAseq data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#the-seurat-object" class="md-nav__link">
    The Seurat object
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/sib-swiss/single-cell-training/edit/master/docs/day1/introduction_scrnaseq.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                  <h1>Introduction scRNAseq</h1>
                
                <h2 id="learning-outcomes">Learning outcomes</h2>
<p><strong>After having completed this chapter you will be able to:</strong></p>
<ul>
<li>Explain what kind of information single cell RNA-seq can give you to answer a biological question</li>
<li>Describe essential considerations during the design of a single cell RNA-seq experiment</li>
<li>Describe the pros and cons of different single cell sequencing methods</li>
<li>Load single cell data into R</li>
<li>Explain the basic structure of a <code>Seurat</code> object and extract count data and metadata</li>
<li>Perform a basic quality control by:<ul>
<li>Evaluating the percentage of UMIs originating from mitochondrial genes</li>
<li>Detecting doublets</li>
</ul>
</li>
</ul>
<h2 id="material">Material</h2>
<p><a class="md-button" href="../../assets/pdf/introduction_scRNAseq.pdf"><span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512"><path d="M181.9 256.1c-5-16-4.9-46.9-2-46.9 8.4 0 7.6 36.9 2 46.9zm-1.7 47.2c-7.7 20.2-17.3 43.3-28.4 62.7 18.3-7 39-17.2 62.9-21.9-12.7-9.6-24.9-23.4-34.5-40.8zM86.1 428.1c0 .8 13.2-5.4 34.9-40.2-6.7 6.3-29.1 24.5-34.9 40.2zM248 160h136v328c0 13.3-10.7 24-24 24H24c-13.3 0-24-10.7-24-24V24C0 10.7 10.7 0 24 0h200v136c0 13.2 10.8 24 24 24zm-8 171.8c-20-12.2-33.3-29-42.7-53.8 4.5-18.5 11.6-46.6 6.2-64.2-4.7-29.4-42.4-26.5-47.8-6.8-5 18.3-.4 44.1 8.1 77-11.6 27.6-28.7 64.6-40.8 85.8-.1 0-.1.1-.2.1-27.1 13.9-73.6 44.5-54.5 68 5.6 6.9 16 10 21.5 10 17.9 0 35.7-18 61.1-61.8 25.8-8.5 54.1-19.1 79-23.2 21.7 11.8 47.1 19.5 64 19.5 29.2 0 31.2-32 19.7-43.4-13.9-13.6-54.3-9.7-73.6-7.2zM377 105 279 7c-4.5-4.5-10.6-7-17-7h-6v128h128v-6.1c0-6.3-2.5-12.4-7-16.9zm-74.1 255.3c4.1-2.7-2.5-11.9-42.8-9 37.1 15.8 42.8 9 42.8 9z"/></svg></span> Download the presentation</a></p>
<ul>
<li>Single cell introductory <a href="https://www.youtube.com/watch?v=k9VFNLLQP8c">video on iBiology</a></li>
<li>Seurat <a href="https://satijalab.org/seurat/">website</a></li>
<li><a href="https://doi.org/10.3389/fcell.2018.00108">Paper</a> on experimental considerations</li>
<li><a href="https://doi.org/10.1093/bib/bby007">Paper</a> on experimental design</li>
<li><a href="https://www.protocols.io/view/smart-seq3-protocol-bcq4ivyw">SMART-seq3 protocol</a> at protocols.io</li>
<li><code>cellranger</code> <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/overview/system-requirements">system requirements</a> and <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/installation">installation</a></li>
<li><a href="https://www.nature.com/articles/s41596-020-00409-w?proof=t">Review</a> by Tallulah Andrews</li>
<li><a href="https://www.biorxiv.org/content/10.1101/749473v3.full">Paper</a> on correlation between mRNA and protein level in single cells</li>
</ul>
<h3 id="running-cellranger-count">Running <code>cellranger count</code></h3>
<p>Have a look in the directory <code>course_data/reads</code> and <code>reference</code>. In the <code>reads</code> directory you will find reads on one sample: <code>ETV6-RUNX1_1</code>. In the analysis part of the course we will work with six samples, but due to time and computational limitations we will run <code>cellranger count</code> on one of the samples, and only reads originating from chromsome 21 and 22. </p>
<p>The only thing you need to run <code>cellranger count</code> are the sequence reads and a reference. Here, we have prepared a reference only with chromosome 21 and 22, but in &lsquo;real life&rsquo; you would of course get the full reference genome of your species. The reference has a specific format. You can download precomputed human and mouse references from the <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest">10X website</a>. If your species of interest is not one of those, you will have to generate it yourself. For that, have a look <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/tutorial_mr">here</a>.</p>
<p>Have a look at the documentation of <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/count"><code>cellranger count</code></a> (scroll down to <em>Command-line argument reference</em>), fill out the missing arguments (at <code>FIXME</code>) in the script below, and run it:</p>
<div class="highlight"><pre><span></span><code>cellranger count <span class="se">\</span>
--id<span class="o">=</span>FIXME <span class="se">\</span>
--sample<span class="o">=</span>FIXME <span class="se">\</span>
--transcriptome<span class="o">=</span>FIXME <span class="se">\</span>
--fastqs<span class="o">=</span>FIXME <span class="se">\</span>
--localcores<span class="o">=</span><span class="m">4</span> 
</code></pre></div>
<div class="admonition note">
<p class="admonition-title">This will take a while..</p>
<p>Once started, the process will need approximately 15 minutes to finish. Have a coffee and/or have a look at the other exercises. </p>
</div>
<div class="admonition hint">
<p class="admonition-title">Running a bash command with Rstudio</p>
<p>You can run a bash script or command using the terminal tab in Rstudio server: 
<figure>
<img src="../../assets/images/select_terminal_tab.png" width="300"/>
</figure></p>
</div>
<details class="done"><summary>Answer</summary><div class="highlight"><pre><span></span><code>cellranger count <span class="se">\</span>
--id<span class="o">=</span>ETV6-RUNX1_1 <span class="se">\</span>
--sample<span class="o">=</span>ETV6-RUNX1_1 <span class="se">\</span>
--transcriptome<span class="o">=</span>cellranger_index <span class="se">\</span>
--fastqs<span class="o">=</span>course_data/reads <span class="se">\</span>
--localcores<span class="o">=</span><span class="m">4</span>
</code></pre></div>
</details>
<p>Have a look out the output directory (i.e. <code>~/ETV6-RUNX1_1/outs</code>). The analysis report (<code>web_summary.html</code>) is usually a good place to start. </p>
<div class="admonition tip">
<p class="admonition-title">Open html files in Rstudio server</p>
<p>You can use the file browser in the bottom right (tab &ldquo;Files&rdquo;) to open html files:</p>
<p><figure>
<img src="../../assets/images/open_html.png" width="300"/>
</figure></p>
</div>
<p><strong>Exercise:</strong> Have a good look inside <code>web_summary.html</code>. Anything that draws your attention? Is this report good enough to continue the analysis? </p>
<details class="done"><summary>Answer</summary><p>Not really. First of all there is a warning: <code>Fraction of RNA read bases with Q-score &gt;= 30 is low</code>. This means that there is a low base quality of the reads. A low base quality gives results in more sequencing error and therefore possibly lower performance while mapping the reads to genes. However, a Q-score of 30 still represents 99.9% accuracy. </p>
<p>What should worry us more is the <strong>number of reads per cell</strong> (363) and the <strong>sequencing saturation</strong> (7.9%). In most cases you should aim for 30.000 - 50.000 reads per cell, and a sequencing saturation &gt;~70% (depending on the application). We therefore don&rsquo;t have enough reads per cell. However, as you might remember, this was a subset of reads (1 million) mapped against chromosome 21 &amp; 22, while the original dataset contains 210,987,037 reads. You can check out the original report at <code>course_data/count_matrices/ETV6-RUNX1_1/outs/web_summary.html</code>.</p>
<p>For more info on sequencing saturation, have a look <a href="https://kb.10xgenomics.com/hc/en-us/articles/115002474263-How-much-sequencing-saturation-should-I-aim-for-">here</a>. </p>
</details>
<h3 id="loading-scrnaseq-data">Loading scRNAseq data</h3>
<p>The next step after the generation of the count matrices with <code>cellranger count</code>, is the data analysis. The <code>R</code> package <code>Seurat</code> is currently the most popular software to do this. To start working with <code>Seurat</code> you can load it into your environment like this:</p>
<div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
</code></pre></div>
<div class="admonition tip">
<p class="admonition-title">Tip: make an R script</p>
<p>You could type and copy-paste the commands of these exercises directly in the console. However, that makes it hard to track what you have done. In addition, it can be nice to add comments to your code, so you can read back why you have made certain choices. In order to do that, do not write commands in the console, but write them in a script, and send them to the console with <span class="keys"><kbd class="key-control">Ctrl</kbd><span>+</span><kbd class="key-enter">Enter</kbd></span> (Windows) or <span class="keys"><kbd class="key-command">Cmd</kbd><span>+</span><kbd class="key-enter">Enter</kbd></span> (MacOS).</p>
</div>
<p>First, we will load a file specifying the different samples, and create an object specifying the location of the count data:</p>
<div class="highlight"><pre><span></span><code><span class="n">sample_info</span> <span class="o">&lt;-</span> <span class="nf">read.csv</span><span class="p">(</span><span class="s">&quot;sample_info_course.csv&quot;</span><span class="p">)</span>
<span class="n">datadirs</span> <span class="o">&lt;-</span> <span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;course_data&quot;</span><span class="p">,</span> <span class="s">&quot;count_matrices&quot;</span><span class="p">,</span> <span class="n">sample_info</span><span class="o">$</span><span class="n">SampleName</span><span class="p">,</span>
                      <span class="s">&quot;outs&quot;</span><span class="p">,</span> <span class="s">&quot;filtered_feature_bc_matrix&quot;</span><span class="p">)</span>
<span class="nf">names</span><span class="p">(</span><span class="n">datadirs</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">gsub</span><span class="p">(</span><span class="s">&quot;_&quot;</span><span class="p">,</span> <span class="s">&quot;-&quot;</span><span class="p">,</span> <span class="n">sample_info</span><span class="o">$</span><span class="n">SampleName</span><span class="p">)</span>
<span class="n">datadirs</span>
</code></pre></div>
<p>The object <code>datadirs</code> is a named vector specifying the paths of the count directories for each sample:</p>
<div class="highlight"><pre><span></span><code>                                                                  PBMMC-1 
     &quot;course_data/count_matrices/PBMMC_1/outs/filtered_feature_bc_matrix&quot; 
                                                                  PBMMC-2 
     &quot;course_data/count_matrices/PBMMC_2/outs/filtered_feature_bc_matrix&quot; 
                                                                  PBMMC-3 
     &quot;course_data/count_matrices/PBMMC_3/outs/filtered_feature_bc_matrix&quot; 
</code></pre></div>
<p>To run through a typical <code>Seurat</code> analysis, we will use the files that are in the directory <code>data/filtered_feature_bc_matrix</code>. This directory is part of the output generated by <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger"><code>cellranger</code></a>. To load this data into R and generate a sparse matrix, run the following command:</p>
<div class="highlight"><pre><span></span><code><span class="n">sparse_matrix</span> <span class="o">&lt;-</span> <span class="n">Seurat</span><span class="o">::</span><span class="nf">Read10X</span><span class="p">(</span><span class="n">data.dir</span> <span class="o">=</span> <span class="n">datadirs</span><span class="p">)</span>
</code></pre></div>
<p>This imports a raw count matrix. Have a look at the counts of the first 30 cells of three genes by running:</p>
<div class="highlight"><pre><span></span><code><span class="n">sparse_matrix</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;PECAM1&quot;</span><span class="p">,</span> <span class="s">&quot;CD8A&quot;</span><span class="p">,</span> <span class="s">&quot;TSPAN1&quot;</span><span class="p">),</span> <span class="m">1</span><span class="o">:</span><span class="m">30</span><span class="p">]</span>
</code></pre></div>
<p>You will see many dots (zeros) and a few integers representing the counts per gene per cell. </p>
<p>To generate a <code>Seurat</code> object, we will run <code>CreateSeuratObject</code>:</p>
<div class="highlight"><pre><span></span><code><span class="n">seu</span> <span class="o">&lt;-</span> <span class="n">Seurat</span><span class="o">::</span><span class="nf">CreateSeuratObject</span><span class="p">(</span><span class="n">counts</span> <span class="o">=</span> <span class="n">sparse_matrix</span><span class="p">,</span>
                                  <span class="n">project</span> <span class="o">=</span> <span class="s">&quot;pbmmc&quot;</span><span class="p">)</span>
</code></pre></div>
<div class="admonition note">
<p class="admonition-title">Function notation with <code>::</code></p>
<p>Here, we define the function together with its associated package. We do that by the syntax <code>package::function</code>. Of course, you can also call <code>library(package)</code>, and only type the function name. Since we use many different packages in this course, it can be confusing which function comes from which package. Therefore, we chose to always associate the package with the called function.</p>
</div>
<p><strong>Exercise:</strong> check what&rsquo;s in the <code>seu</code> object, by typing <code>seu</code> in the R console. How many features are in there? And how many cells?</p>
<details class="done"><summary>Answer</summary><p>Typing <code>seu</code> should return:</p>
<div class="highlight"><pre><span></span><code>An object of class Seurat 
18673 features across 6946 samples within 1 assay 
Active assay: RNA (18673 features, 0 variable features)
</code></pre></div>
<p>This means that we have 18673 genes (features) in there, and 6946 cells (samples)</p>
</details>
<h3 id="the-seurat-object">The <code>Seurat</code> object</h3>
<p>The <code>seu</code> object we have created has the class <code>Seurat</code>. The object contains multi-level slots and lists. Each <code>Seurat</code> object contains exactly the same slots that are specified in the image below (get slot descriptions by typing <code>?SeuratObject::Seurat</code>). You can get the information inside a slot with <code>@</code>, in the same way as you would use the <code>$</code> for lists (e.g. <code>seu@meta.data</code> will return the a data frame with information on each cell). Slots can be filled with other R objects, like lists, vectors, data frames or any other class. Here&rsquo;s an overview of all slots that are in a <code>Seurat</code> object:</p>
<figure>
  <img src="../../assets/images/seurat_object.png" width="500"/>
</figure>

<p>In addition to the original count table, the <code>Seurat</code> object can therefore store a lot of information that is generated during your analysis, like results of a normalization (<code>@assays$RNA@data</code>) a PCA or UMAP (<code>@reductions</code>) and the clustering (<code>@graphs</code>). It also tracks all the commands that have been used to generate the object in its current state (<code>@commands</code>). Therefore, while going through the analysis steps, the same object gets more and more of its slots filled. Because most <code>Seurat</code> functions return the input object + adjusted slots, we can use this syntax:</p>
<div class="highlight"><pre><span></span><code><span class="n">seurat_object</span> <span class="o">&lt;-</span> <span class="n">Seurat</span><span class="o">::</span><span class="nf">function</span><span class="p">(</span><span class="n">seurat_object</span><span class="p">)</span>
</code></pre></div>
<p>So, the function takes an object as input and we assign it to an object with the same name. Meaning that we overwrite the object used as input. This is fine in many cases, because <code>Seurat</code> adds information to the input object, and returns is. We will use this syntax for the first time at <a href="../normalization_scaling/">Normalization and scaling</a>.</p>
<div class="admonition note">
<p class="admonition-title">Getting specific information out of the <code>Seurat</code> object<p>In order to get specific data you can use the <code>@</code> and <code>$</code> symbols to browse through the objects. However, <code>Seurat</code> comes with a lot of convenience functions, that are easier to use. So, e.g. to get the raw count matrix, you could type <code>seurat_object@assays$RNA@counts</code>, however, this is equivalent to <code>GetAssayData(object = seurat_object, slot = "counts")</code>. More information on these convenience functions <a href="https://satijalab.org/seurat/articles/essential_commands.html">here</a>.</p>
</p>
</div>
<p><strong>Exercise:</strong></p>
<p>A. Have a look at the <code>seu</code> object by running <code>View(seu)</code>. What is in there? What is stored in <code>@active.ident</code>? </p>
<p>B. Have a look at the <code>data.frame</code> stored at <code>seu@meta.data</code> what kind of information is in there?</p>
<details class="done"><summary>Answer</summary><div class="tabbed-set" data-tabs="1:2"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><label for="__tabbed_1_1">Answer A</label><div class="tabbed-content">
<p>There are many slots as described in the above figure. The slot <code>@active.ident</code> contains data specifying the samples, e.g. <code>table(seu@active.ident)</code> returns:</p>
<div class="highlight"><pre><span></span><code>PBMMC-1 PBMMC-2 PBMMC-3 
1612    3105    2229 
</code></pre></div>
<p>Which are the number of cells per sample. </p>
</div>
<input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><label for="__tabbed_1_2">Answer B</label><div class="tabbed-content">
<p>Running <code>head(seu@meta.data)</code> returns:</p>
<div class="highlight"><pre><span></span><code>                           orig.ident nCount_RNA nFeature_RNA
PBMMC-1_AAACCTGCAGACGCAA-1    PBMMC-1       2401          909
PBMMC-1_AAACCTGTCATCACCC-1    PBMMC-1       3532          760
PBMMC-1_AAAGATGCATAAAGGT-1    PBMMC-1       3972         1215
PBMMC-1_AAAGCAAAGCAGCGTA-1    PBMMC-1       3569          894
PBMMC-1_AAAGCAACAATAACGA-1    PBMMC-1       2982          730
PBMMC-1_AAAGCAACATCAGTCA-1    PBMMC-1      22284         3108
</code></pre></div>
<p>Giving you the names of three columns and a row for each cell:</p>
<ul>
<li><code>orig_ident</code>: the original identity (origin) of a cell.</li>
<li><code>nCount_RNA</code>: the number of reads assigned to a cell.</li>
<li><code>nFeature_RNA</code>: the number of expressed features (genes) per cell.</li>
</ul>
</div>
</div>
</details>
<p>Luckily, usually you do not have to dive into this structure to retrieve information. For example, information in the slot <code>@meta.data</code> can be retrieved and set by using <code>$</code> or <code>[[]]</code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>There is a subtle difference here between <code>$</code> and <code>[[]]</code>. While <code>$</code> returns a vector of the column in <code>@meta.data</code>, <code>[[]]</code> returns a <code>data.frame</code>.</p>
</div>
<p><strong>Exercise:</strong> Generate a histogram of the column <code>nCount_RNA</code> at <code>seu@meta.data</code>, with the base function <code>hist</code>.</p>
<details class="done"><summary>Answer</summary><div class="highlight"><pre><span></span><code><span class="nf">hist</span><span class="p">(</span><span class="n">seu</span><span class="o">$</span><span class="n">nCount_RNA</span><span class="p">)</span>
</code></pre></div>
<p>or</p>
<div class="highlight"><pre><span></span><code><span class="nf">hist</span><span class="p">(</span><span class="n">seu</span><span class="o">@</span><span class="n">meta.data</span><span class="o">$</span><span class="n">nCount_RNA</span><span class="p">)</span>
</code></pre></div>
</details>
<p>There are also built-in functions to plot data from <code>Seurat</code> object, for example <code>FeatureScatter</code>. This function enables you easily draw a scatterplot from a <code>Seurat</code> object:</p>
<div class="highlight"><pre><span></span><code><span class="n">Seurat</span><span class="o">::</span><span class="nf">FeatureScatter</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">feature1</span> <span class="o">=</span> <span class="s">&quot;nCount_RNA&quot;</span><span class="p">,</span> <span class="n">feature2</span> <span class="o">=</span> <span class="s">&quot;nFeature_RNA&quot;</span><span class="p">)</span>
</code></pre></div>
<p>You can find a nice overview of such functions <a href="https://satijalab.org/seurat/articles/essential_commands.html">here</a>. </p>
                
              
              
                


              
            </article>
          </div>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../setup/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Setup" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Setup
            </div>
          </div>
        </a>
      
      
        
        <a href="../quality_control/" class="md-footer__link md-footer__link--next" aria-label="Next: Quality control" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Quality control
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../..", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../../assets/javascripts/workers/search.409db549.min.js", "version": {"provider": "mike"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.56a63758.min.js"></script>
      
    
  </body>
</html>