
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      <link rel="shortcut icon" href="../../assets/images/SIB_logo.svg">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-6.2.8">
    
    
      
        <title>Normalization and scaling - Single cell transcriptomics</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.cb6bc1d0.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.39b8e14a.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto",-apple-system,BlinkMacSystemFont,Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono",SFMono-Regular,Consolas,Menlo,monospace}</style>
      
    
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
      
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#learning-outcomes" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid" aria-label="Header">
    <a href="../.." title="Single cell transcriptomics" class="md-header-nav__button md-logo" aria-label="Single cell transcriptomics">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    <label class="md-header-nav__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header-nav__title" data-md-component="header-title">
      <div class="md-header-nav__ellipsis">
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            Single cell transcriptomics
          </span>
        </div>
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            
              Normalization and scaling
            
          </span>
        </div>
      </div>
    </div>
    
      <label class="md-header-nav__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0116 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 019.5 16 6.5 6.5 0 013 9.5 6.5 6.5 0 019.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0116 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 019.5 16 6.5 6.5 0 013 9.5 6.5 6.5 0 019.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" data-md-component="search-reset" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header-nav__source">
        
<a href="https://github.com/sib-swiss/single-cell-training/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/single-cell-training
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    




<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Single cell transcriptomics" class="md-nav__button md-logo" aria-label="Single cell transcriptomics">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    Single cell transcriptomics
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/sib-swiss/single-cell-training/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/single-cell-training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../precourse/" class="md-nav__link">
        Precourse preparations
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../course_schedule/" class="md-nav__link">
        Course schedule
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-4" type="checkbox" id="nav-4" checked>
      
      <label class="md-nav__link" for="nav-4">
        Day 1
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Day 1" data-md-level="1">
        <label class="md-nav__title" for="nav-4">
          <span class="md-nav__icon md-icon"></span>
          Day 1
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../general_introduction/" class="md-nav__link">
        General introduction
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../introduction_scrnaseq/" class="md-nav__link">
        Introduction scRNAseq
      </a>
    </li>
  

          
            
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Normalization and scaling
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Normalization and scaling
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    Learning outcomes
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    Material
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    Exercises
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#normalization" class="md-nav__link">
    Normalization
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#variable-features" class="md-nav__link">
    Variable features
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scaling" class="md-nav__link">
    Scaling
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../quality_control/" class="md-nav__link">
        Quality control
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-5" type="checkbox" id="nav-5" >
      
      <label class="md-nav__link" for="nav-5">
        Day 2
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Day 2" data-md-level="1">
        <label class="md-nav__title" for="nav-5">
          <span class="md-nav__icon md-icon"></span>
          Day 2
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/dimensionality_reduction/" class="md-nav__link">
        Dimensionality reduction
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/integration/" class="md-nav__link">
        Integration
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/cell_annotation/" class="md-nav__link">
        Cell annotation
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-6" type="checkbox" id="nav-6" >
      
      <label class="md-nav__link" for="nav-6">
        Day 3
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Day 3" data-md-level="1">
        <label class="md-nav__title" for="nav-6">
          <span class="md-nav__icon md-icon"></span>
          Day 3
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../day3/differential_gene_expression/" class="md-nav__link">
        Differential gene expression
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../day3/enrichment_analysis/" class="md-nav__link">
        Erichment analysis
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../day3/trajectory_analysis/" class="md-nav__link">
        Trajectory analysis
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    Learning outcomes
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    Material
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    Exercises
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#normalization" class="md-nav__link">
    Normalization
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#variable-features" class="md-nav__link">
    Variable features
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scaling" class="md-nav__link">
    Scaling
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/sib-swiss/single-cell-training/edit/master/docs/day1/normalization_scaling.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                  <h1>Normalization and scaling</h1>
                
                <h2 id="learning-outcomes">Learning outcomes</h2>
<p><strong>After having completed this chapter you will be able to:</strong></p>
<ul>
<li>Describe and perform standard procedures for normalization and scaling with the package <code>Seurat</code></li>
<li>Select the most variable genes from a <code>Seurat</code> object for downstream analyses</li>
</ul>
<h2 id="material">Material</h2>
<ul>
<li><a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html">Seurat vignette</a></li>
</ul>
<h2 id="exercises">Exercises</h2>
<blockquote>
<p><span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M6.1 444.3c-9.6 10.8-7.5 27.6 4.5 35.7l68.8 27.9c9.9 6.7 23.3 5 31.3-3.8l91.8-101.9-79.2-87.9-117.2 130zm435.8 0s-292-324.6-295.4-330.1c15.4-8.4 40.2-17.9 77.5-17.9s62.1 9.5 77.5 17.9c-3.3 5.6-56 64.6-56 64.6l79.1 87.7 34.2-38c28.7-31.9 33.3-78.6 11.4-115.5l-43.7-73.5c-4.3-7.2-9.9-13.3-16.8-18-40.7-27.6-127.4-29.7-171.4 0-6.9 4.7-12.5 10.8-16.8 18l-43.6 73.2c-1.5 2.5-37.1 62.2 11.5 116L337.5 504c8 8.9 21.4 10.5 31.3 3.8l68.8-27.9c11.9-8 14-24.8 4.3-35.6z"/></svg></span> This chapter uses the <code>gbm</code> dataset</p>
</blockquote>
<h3 id="normalization">Normalization</h3>
<div class="admonition bug">
<p class="admonition-title">Bug</p>
<p>Recommendations are unclear here. Should we use the <code>NormalizeData</code> -&gt; <code>FindVariableFeatures</code> -&gt; <code>ScaleData</code> workflow or <code>SCTransform</code>??</p>
</div>
<p>After removing unwanted cells from the dataset, the next step is to normalize the data.
By default, Seurat employs a global-scaling normalization method <code>"LogNormalize"</code> that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result.
Normalized values are stored in the &ldquo;RNA&rdquo; slot of the gbm object.</p>
<p><strong>Exercise:</strong> Have a look at the assay data before and after running <code>NormalizeData()</code>. Did it change?</p>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>You can extract assay data with the function <code>Seurat::GetAssay</code></p>
</div>
<details class="done"><summary>Answer</summary><p>You can check out some assay data with:</p>
<p><div class="highlight"><pre><span></span><code><span class="n">Seurat</span><span class="o">::</span><span class="nf">GetAssay</span><span class="p">(</span><span class="n">gbm</span><span class="p">)</span><span class="n">[1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="n">]</span>  
</code></pre></div>
Returning:</p>
<div class="tabbed-set" data-tabs="1:2"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><label for="__tabbed_1_1">Before normalization</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code>10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
  [[ suppressing 10 column names ‘AAACCCAAGGCGATAC-1’, ‘AAACCCACAAGTCCCG-1’, ‘AAACCCACAGATGCGA-1’ ... ]]

AL627309.1 . . . . . . . . . .
AL627309.5 . . . . . . . . . .
AP006222.2 . . . . . . . . . .
LINC01409  . . . 1 . . . . . .
FAM87B     . . . . . . . . . .
LINC01128  . . 1 . . . . 1 . 1
LINC00115  . . . . . . . . . .
FAM41C     . . . . . . . . . .
AL645608.6 . . . . . . . . . .
AL645608.2 . . . . . . . . . .
</code></pre></div>

</div>
<input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><label for="__tabbed_1_2">After normalization</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code>10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
   [[ suppressing 10 column names ‘AAACCCAAGGCGATAC-1’, ‘AAACCCACAAGTCCCG-1’, ‘AAACCCACAGATGCGA-1’ ... ]]

AL627309.1 . . .         .         . . . .         . .        
AL627309.5 . . .         .         . . . .         . .        
AP006222.2 . . .         .         . . . .         . .        
LINC01409  . . .         0.7438965 . . . .         . .        
FAM87B     . . .         .         . . . .         . .        
LINC01128  . . 0.7991683 .         . . . 0.5091777 . 0.3826447
LINC00115  . . .         .         . . . .         . .        
FAM41C     . . .         .         . . . .         . .        
AL645608.6 . . .         .         . . . .         . .        
AL645608.2 . . .         .         . . . .         . .  
</code></pre></div>

</div>
</div>
</details>
<div class="highlight"><pre><span></span><code><span class="n">gbm</span> <span class="o">&lt;-</span> <span class="n">Seurat</span><span class="o">::</span><span class="nf">NormalizeData</span><span class="p">(</span><span class="n">gbm</span><span class="p">,</span>
                     <span class="n">normalization.method</span> <span class="o">=</span> <span class="s">&quot;LogNormalize&quot;</span><span class="p">,</span>
                     <span class="n">scale.factor</span> <span class="o">=</span> <span class="m">10000</span><span class="p">)</span>
</code></pre></div>

<div class="admonition note">
<p class="admonition-title">Updating <code>gbm</code></p>
<p>As you might have noticed, this function takes the object <code>gbm</code> as input, and it returns it to an object named <code>gbm</code>. We can do this because the output of such calculations are added to the object, without loosing information.</p>
</div>
<h3 id="variable-features">Variable features</h3>
<p>We next calculate a subset of features that exhibit high cell-to-cell variation in the
dataset (i.e, they are highly expressed in some cells, and lowly expressed in others).
Focusing on these genes in downstream analysis helps to highlight biological signal
in single-cell datasets.
The procedure in Seurat models the mean-variance relationship inherent in single-cell
data, and is implemented in the <code>FindVariableFeatures()</code> function.
By default, 2,000 genes (features) per dataset are returned and these will be used in
downstream analysis, like PCA.</p>
<div class="highlight"><pre><span></span><code><span class="n">gbm</span> <span class="o">&lt;-</span> <span class="n">Seurat</span><span class="o">::</span><span class="nf">FindVariableFeatures</span><span class="p">(</span><span class="n">gbm</span><span class="p">,</span>
                            <span class="n">selection.method</span> <span class="o">=</span> <span class="s">&quot;vst&quot;</span><span class="p">,</span>
                            <span class="n">nfeatures</span> <span class="o">=</span> <span class="m">2000</span><span class="p">)</span>
</code></pre></div>

<p>Let&rsquo;s have a look at the 10 most variable genes:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Identify the 10 most highly variable genes</span>
<span class="n">top10</span> <span class="o">&lt;-</span> <span class="nf">head</span><span class="p">(</span><span class="n">Seurat</span><span class="o">::</span><span class="nf">VariableFeatures</span><span class="p">(</span><span class="n">gbm</span><span class="p">),</span> <span class="m">10</span><span class="p">)</span>
<span class="n">top10</span>
</code></pre></div>

<p>We can plot them in a nicely labeled scatterplot:</p>
<div class="highlight"><pre><span></span><code><span class="n">vf_plot</span> <span class="o">&lt;-</span> <span class="n">Seurat</span><span class="o">::</span><span class="nf">VariableFeaturePlot</span><span class="p">(</span><span class="n">gbm</span><span class="p">)</span>
<span class="n">Seurat</span><span class="o">::</span><span class="nf">LabelPoints</span><span class="p">(</span><span class="n">plot</span> <span class="o">=</span> <span class="n">vf_plot</span><span class="p">,</span>
            <span class="n">points</span> <span class="o">=</span> <span class="n">top10</span><span class="p">,</span> <span class="n">repel</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>

<h3 id="scaling">Scaling</h3>
<p>Next, we apply scaling, a linear transformation that is a standard pre-processing
step prior to dimensional reduction techniques like PCA. The <code>ScaleData()</code> function</p>
<ol>
<li>shifts the expression of each gene, so that the mean expression across cells is 0</li>
<li>scales the expression of each gene, so that the variance across cells is 1</li>
</ol>
<p>This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate. The results of this are stored in <code>gbm$RNA@scale.data</code></p>
<div class="highlight"><pre><span></span><code><span class="n">gbm</span> <span class="o">&lt;-</span> <span class="n">Seurat</span><span class="o">::</span><span class="nf">ScaleData</span><span class="p">(</span><span class="n">gbm</span><span class="p">,</span>
                 <span class="n">features</span> <span class="o">=</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">gbm</span><span class="p">))</span>
</code></pre></div>

<p>Here&rsquo;s a function to plot mean versus variance in a count matrix (credits go to Wandrille Duchemin):</p>
<div class="highlight"><pre><span></span><code><span class="n">plot_mean_variance</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">count_mat</span><span class="p">){</span>
  <span class="c1"># get log10 of mean per gene</span>
  <span class="n">log_mean</span> <span class="o">&lt;-</span> <span class="nf">log10</span><span class="p">(</span><span class="nf">rowMeans</span><span class="p">(</span><span class="n">count_mat</span><span class="p">))</span>
  <span class="c1"># get log10 of variance per gene</span>
  <span class="n">log_var</span> <span class="o">&lt;-</span> <span class="nf">log10</span><span class="p">(</span><span class="nf">apply</span><span class="p">(</span><span class="n">count_mat</span><span class="p">,</span> <span class="m">1</span><span class="p">,</span> <span class="n">var</span><span class="p">))</span>
  <span class="c1"># get density colors</span>
  <span class="n">cols</span> <span class="o">&lt;-</span> <span class="nf">densCols</span><span class="p">(</span><span class="n">log_mean</span><span class="p">,</span> <span class="n">log_var</span><span class="p">)</span>

  <span class="nf">plot</span><span class="p">(</span><span class="n">log_mean</span><span class="p">,</span> <span class="n">log_var</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">cols</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">,</span> <span class="n">cex</span> <span class="o">=</span> <span class="m">0.2</span><span class="p">,</span>
        <span class="n">xlab</span> <span class="o">=</span> <span class="s">&quot;log10(mean)&quot;</span><span class="p">,</span> <span class="n">ylab</span> <span class="o">=</span> <span class="s">&quot;log10(variation)&quot;</span><span class="p">)</span>
  <span class="c1"># draw the line y=x</span>
  <span class="nf">abline</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="s">&quot;red&quot;</span><span class="p">)</span>
<span class="p">}</span>
</code></pre></div>

<p><strong>Exercise:</strong> Use the above function to plot the mean variance of:</p>
<ul>
<li>raw counts (stored in the slot <code>"counts"</code>)</li>
<li>normalized data (stored in the slot <code>"data"</code>)</li>
<li>scaled data (stored in the slot <code>"scale.data"</code>)</li>
</ul>
<div class="admonition hint">
<p class="admonition-title">Retrieving count data</p>
<p>You can retrieve count data from a specific slot with the function <code>GetAssayData</code>, e.g. for the raw counts this would be:
<div class="highlight"><pre><span></span><code><span class="n">Seurat</span><span class="o">::</span><span class="nf">GetAssayData</span><span class="p">(</span><span class="n">gbm</span><span class="p">,</span> <span class="n">slot</span> <span class="o">=</span> <span class="s">&quot;counts&quot;</span><span class="p">)</span>
</code></pre></div></p>
</div>
<details class="done"><summary>Answer</summary><div class="tabbed-set" data-tabs="2:3"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><label for="__tabbed_2_1">Raw counts</label><div class="tabbed-content">
<p>Running</p>
<div class="highlight"><pre><span></span><code><span class="nf">plot_mean_variance</span><span class="p">(</span><span class="n">Seurat</span><span class="o">::</span><span class="nf">GetAssayData</span><span class="p">(</span><span class="n">gbm</span><span class="p">,</span> <span class="n">slot</span> <span class="o">=</span>  <span class="s">&quot;counts&quot;</span><span class="p">))</span>
</code></pre></div>

<p>Returns:</p>
<p><figure>
  <img src="../../assets/images/plot_mean_var_counts.png" width="400"/>
</figure></p>
</div>
<input id="__tabbed_2_2" name="__tabbed_2" type="radio" /><label for="__tabbed_2_2">Normalized data</label><div class="tabbed-content">
<p>Running</p>
<div class="highlight"><pre><span></span><code><span class="nf">plot_mean_variance</span><span class="p">(</span><span class="n">Seurat</span><span class="o">::</span><span class="nf">GetAssayData</span><span class="p">(</span><span class="n">gbm</span><span class="p">,</span> <span class="n">slot</span> <span class="o">=</span>  <span class="s">&quot;data&quot;</span><span class="p">))</span>
</code></pre></div>

<p>Returns:</p>
<p><figure>
  <img src="../../assets/images/plot_mean_var_data.png" width="400"/>
</figure></p>
</div>
<input id="__tabbed_2_3" name="__tabbed_2" type="radio" /><label for="__tabbed_2_3">Scaled and normalized</label><div class="tabbed-content">
<p>Running</p>
<div class="highlight"><pre><span></span><code><span class="nf">plot_mean_variance</span><span class="p">(</span><span class="n">Seurat</span><span class="o">::</span><span class="nf">GetAssayData</span><span class="p">(</span><span class="n">gbm</span><span class="p">,</span> <span class="n">slot</span> <span class="o">=</span>  <span class="s">&quot;scale.data&quot;</span><span class="p">))</span>
</code></pre></div>

<p>As you remember, scaling results in a variance of 1 for each gene, so the <code>log10</code> of the variance would be 0. In addition, the mean is scaled to zero, so a <code>log10</code> of 0 would return <code>-Inf</code> or a very negative number (in case it&rsquo;s very close to 0). It therefore returns:</p>
<p><figure>
  <img src="../../assets/images/plot_mean_var_scale.data.png" width="400"/>
</figure></p>
</div>
</div>
</details>
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid" aria-label="Footer">
        
          <a href="../introduction_scrnaseq/" class="md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
            </div>
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Introduction scRNAseq
              </div>
            </div>
          </a>
        
        
          <a href="../quality_control/" class="md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Quality control
              </div>
            </div>
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../assets/javascripts/vendor.18f0862e.min.js"></script>
      <script src="../../assets/javascripts/bundle.994580cf.min.js"></script><script id="__lang" type="application/json">{"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}</script>
      
      <script>
        app = initialize({
          base: "../..",
          features: [],
          search: Object.assign({
            worker: "../../assets/javascripts/worker/search.9c0e82ba.min.js"
          }, typeof search !== "undefined" && search)
        })
      </script>
      
    
  </body>
</html>